core/src/main/java/io/questdb/cairo/mv/MatViewRefreshJob.java

```diff
@@ -178,62 +180,152 @@ private void enqueueInvalidateDependentViews(TableToken viewToken, String invali
+        final boolean rangeRefresh = rangeTo != Numbers.LONG_NULL;
+        final boolean incrementalRefresh = lastRefreshTxn != Numbers.LONG_NULL;
+
         LongList txnIntervals = null;
-        long minTs;
-        long maxTs;
-        if (lastRefreshTxn > 0) {
-            // Incremental refresh.
-            // Find min/max timestamps from WAL transactions.
+        long minTs = Long.MAX_VALUE;
+        long maxTs = Long.MIN_VALUE;
+        if (incrementalRefresh) {
+            // Incremental refresh. This means that there may be a data transaction in the base table
+            // or someone has run REFRESH INCREMENTAL SQL.
+            // Let's find min/max timestamps in the new WAL transactions.
             txnIntervals = intervals;
             txnIntervals.clear();
-            txnRangeLoader.load(engine, Path.PATH.get(), baseTableToken, txnIntervals, lastRefreshTxn, lastTxn);
-            minTs = txnRangeLoader.getMinTimestamp();
-            maxTs = txnRangeLoader.getMaxTimestamp();
-            // Check if refresh limit should be applied.
-            try (TableMetadata matViewMeta = engine.getTableMetadata(viewToken)) {
-                final int limit = matViewMeta.getMatViewRefreshLimitHoursOrMonths();
-                if (limit != 0) {
-                    final long now = microsecondClock.getTicks();
-                    if (limit > 0) { // hours
-                        minTs = Math.max(minTs, now - Timestamps.HOUR_MICROS * limit);
-                    } else { // months
-                        minTs = Math.max(minTs, Timestamps.addMonths(now, -limit));
-                    }
+            if (lastRefreshTxn > -1) {
+                // It's a subsequent incremental refresh, so WalPurgeJob must be aware of us.
+                txnRangeLoader.load(engine, Path.PATH.get(), baseTableToken, txnIntervals, lastRefreshTxn, lastTxn);
+                minTs = txnRangeLoader.getMinTimestamp();
+                maxTs = txnRangeLoader.getMaxTimestamp();
+            } else {
+                // It's the first incremental refresh. WAL segments may be already purged,
+                // so let's take min/max timestamps from the reader.
+                minTs = baseTableReader.getMinTimestamp();
+                maxTs = baseTableReader.getMaxTimestamp();
+                txnIntervals.add(minTs, maxTs);
+            }
+        } else if (rangeRefresh) {
+            // Range refresh. This means that the timer is triggered on a period mat view
+            // or someone has run REFRESH RANGE SQL.
+            // In both cases we have the range to be refreshed.
+            if (rangeFrom == Numbers.LONG_NULL) {
+                // Period timer has triggered.
+                long periodLo = viewState.getLastPeriodHi();
+                if (periodLo == Numbers.LONG_NULL) {
+                    periodLo = baseTableReader.getMinTimestamp();
+                }
+                if (periodLo < rangeTo) {
```

Comment:
I think it should be
```suggestion
                if (periodLo <= rangeTo) {
```